### ISSUE TITLE
runtime: A QEMU signaling fix causes Go binaries to fail on startup.

### ISSUE THREAD

**Issue by joshkunz** [golang/go#33746]
tl;dr: In anticipation of a future QEMU work-around, Go should hide runtime `rt_sigaction` failures for `SIGRTMAX - 1`, like [it already does](https://github.com/golang/go/blob/master/src/runtime/os_linux.go#L432-L448) for `SIGRTMAX`.

QEMU user mode emulation has [a bug](https://bugs.launchpad.net/qemu/+bug/1594394) where it will not correctly deliver the `SIGRTMIN + 1` signal. This is because `SIGRTMIN + 1` is a glibc reserved signal, and it is caught in the emulator's glibc. To get around this issue there is [a patch](http://patchwork.ozlabs.org/patch/590640/) to map `SIGRTMIN + 1` to `SIGRTMAX - 1`, a signal that is typically unused. QEMU already has a similar work-around, mapping `SIGRTMIN` to ` SIGRTMAX` which is also a signal reserved by glibc.

On startup, the Go attempts to register signal handlers for all signals, but it silently hides failures for `SIGRTMIN`, `SIGRTMIN + 1` (the glibc reserved signals), and `SIGRTMAX` (the QEMU mapped signal for `SIGRTMIN`). If the QEMU patch is submitted, go programs running under QEMU will fail to run, due to panics when trying to register a handler for `SIGRTMAX - 1`.

If Go decides to preemptively fix this issue, the patch is simple: just ignore failures for signal `63` in addition to `64`. I will follow up this issue with a patch shortly.

**Comment by joshkunz** [comment]
@dsnet who I mentioned this to previously.

**Comment by dsnet** [comment]
\cc @ianlancetaylor

**Comment by bmkessler** [comment]
Is there a more robust way to handle this issue?  From the signal(7) manpage http://man7.org/linux/man-pages/man7/signal.7.html:

       Real-time signals
       Starting with version 2.2, Linux supports real-time signals as
       originally defined in the POSIX.1b real-time extensions (and now
       included in POSIX.1-2001).  The range of supported real-time signals
       is defined by the macros SIGRTMIN and SIGRTMAX.  POSIX.1-2001
       requires that an implementation support at least _POSIX_RTSIG_MAX (8)
       real-time signals.

       The Linux kernel supports a range of 33 different real-time signals,
       numbered 32 to 64.  However, the glibc POSIX threads implementation
       internally uses two (for NPTL) or three (for LinuxThreads) real-time
       signals (see pthreads(7)), and adjusts the value of SIGRTMIN suitably
       (to 34 or 35).  Because the range of available real-time signals
       varies according to the glibc threading implementation (and this
       variation can occur at run time according to the available kernel and
       glibc), and indeed the range of real-time signals varies across UNIX
       systems, programs should never refer to real-time signals using hard-
       coded numbers, but instead should always refer to real-time signals
       using the notation SIGRTMIN+n, and include suitable (run-time) checks
       that SIGRTMIN+n does not exceed SIGRTMAX.

Which says values 63 and 64 that qemu ends up using in the common case when SIGRTMAX==64 shouldn't be hardcoded.

I'm bringing it up because currently Go programs compiled for linux/mips (mips mipsle mips64 mips64le) fail on startup running in qemu user mode on most architectures (amd64 etc.)  due to a related issue. The go compiler hardcodes that realtime signals go up to 128 on linux/mips (implicitly that SIGRTMAX==128 on linux/mips  https://github.com/golang/go/blob/77f9b2728eb08456899e6500328e00ec4829dddf/src/runtime/sigtab_linux_mipsx.go) and on startup sigaction on 65 and above fail on amd64 etc.  It would be nice to have a solution general enough to allow mips programs to run under qemu user mode as well.

**Comment by ianlancetaylor** [comment]
From further discussion the situation with QEMU is complex.  This issue was filed in anticipation of changes that have not yet occurred.  So I am putting the issue on hold for now.

**Comment by tomkcook** [comment]
Unsure if it's related, but I see similar crashes in the go compiler (ie `go build`) for both go 1.11 and go 1.13 running on any version of qemu (I've tried most releases from 3.0.0 to 4.1.0) when building a "Hello, world" program.

**Comment by ianlancetaylor** [comment]
@tomkcook Can you clarify what you mean by "similar crashes?"  Can you show us the output of the crash?  Thanks.

**Comment by bcmills** [comment]
Note that we currently lack a QEMU builder (#1508), and per http://golang.org/wiki/PortingPolicy, supported platforms require builders.

I don't think we should be putting in QEMU-specific workarounds until/unless we have a builder to test for regressions in those workarounds.

**Comment by yuchunyu97** [comment]
When I use qemu user mode on X86 platform to run go binary files on MIPS platform, I got the following error:

```bash
fatal error: sigaction failed

runtime stack:
runtime.throw(0x123cc73, 0x10)
	/usr/local/go/src/runtime/panic.go:617 +0x6c
runtime.sysSigaction.func1()
	/usr/local/go/src/runtime/os_linux.go:432 +0x4c
runtime.sysSigaction(0xc000000041, 0x4000800488, 0x0)
	/usr/local/go/src/runtime/os_linux.go:431 +0x98
runtime.sigaction(0x41, 0x4000800488, 0x0)
	/usr/local/go/src/runtime/sigaction.go:15 +0x28
runtime.setsig(0x41, 0x607f0)
	/usr/local/go/src/runtime/os_linux.go:382 +0xd8
runtime.initsig(0x1f300)
	/usr/local/go/src/runtime/signal_unix.go:113 +0x288
runtime.mstartm0()
	/usr/local/go/src/runtime/proc.go:1230 +0x60
runtime.mstart1()
	/usr/local/go/src/runtime/proc.go:1202 +0x118
runtime.mstart()
	/usr/local/go/src/runtime/proc.go:1172 +0x6c

goroutine 1 [runnable]:
runtime.main()
	/usr/local/go/src/runtime/proc.go:110
runtime.goexit()
	/usr/local/go/src/runtime/asm_mips64x.s:646 +0x4
```

Can anyone help me?

**Comment by yuchunyu97** [comment]
> When I use qemu user mode on X86 platform to run go binary files on MIPS platform, I got the following error:
> 
> ```shell
> fatal error: sigaction failed
> 
> runtime stack:
> runtime.throw(0x123cc73, 0x10)
> 	/usr/local/go/src/runtime/panic.go:617 +0x6c
> runtime.sysSigaction.func1()
> 	/usr/local/go/src/runtime/os_linux.go:432 +0x4c
> runtime.sysSigaction(0xc000000041, 0x4000800488, 0x0)
> 	/usr/local/go/src/runtime/os_linux.go:431 +0x98
> runtime.sigaction(0x41, 0x4000800488, 0x0)
> 	/usr/local/go/src/runtime/sigaction.go:15 +0x28
> runtime.setsig(0x41, 0x607f0)
> 	/usr/local/go/src/runtime/os_linux.go:382 +0xd8
> runtime.initsig(0x1f300)
> 	/usr/local/go/src/runtime/signal_unix.go:113 +0x288
> runtime.mstartm0()
> 	/usr/local/go/src/runtime/proc.go:1230 +0x60
> runtime.mstart1()
> 	/usr/local/go/src/runtime/proc.go:1202 +0x118
> runtime.mstart()
> 	/usr/local/go/src/runtime/proc.go:1172 +0x6c
> 
> goroutine 1 [runnable]:
> runtime.main()
> 	/usr/local/go/src/runtime/proc.go:110
> runtime.goexit()
> 	/usr/local/go/src/runtime/asm_mips64x.s:646 +0x4
> ```
> 
> Can anyone help me?

This problem occurs with semaphores at and after 65

**Comment by joshkunz** [comment]
@yuchunyu97 QEMU developers are likely not following this forum closely. To report QEMU issues you can use [QEMU's Launchpad](https://bugs.launchpad.net/qemu) page instead.

Just a hint for your issue: A [patch was recently merged](https://github.com/qemu/qemu/commit/6bc024e713fd35eb5fddbe16acd8dc92d27872a9) which could fix this issue. Consider trying with a QEMU built from the master branch.

**Comment by yuchunyu97** [comment]
> @yuchunyu97 QEMU developers are likely not following this forum closely. To report QEMU issues you can use [QEMU's Launchpad](https://bugs.launchpad.net/qemu) page instead.
> 
> Just a hint for your issue: A [patch was recently merged](https://github.com/qemu/qemu/commit/6bc024e713fd35eb5fddbe16acd8dc92d27872a9) which could fix this issue. Consider trying with a QEMU built from the master branch.

@joshkunz 
Thank you for your help, the latest qemu does work.
I just want to run MIPS go commands normally on X86 platforms, but after solving the previous problem, a new problem appeared, which does not look like the problem of qemu.
Do you know this?

```bash
$ go version
panic: regexp: Compile(`(?i)^[[:space:]]*(unordered )?output:`): error parsing regexp: invalid or unsupported Perl syntax: `(?i`

goroutine 1 [running]:
regexp.MustCompile(0x660a49, 0x25, 0xc00007e2a0)
	/root/tmp/go-linux-mips64le-bootstrap/src/regexp/regexp.go:272 +0x158
go/doc.init.ializers()
	/root/tmp/go-linux-mips64le-bootstrap/src/go/doc/example.go:107 +0x22c
```

**Comment by bradfitz** [comment]
Not sure when it happened, but I noticed that qemu-user 5.2.0 in Debian Bullseye works with Go binaries!

As one bisect datapoint, qemu 4.2 in Ubuntu Focal doesn't work.

(in this case, I tested Go 1.17 binaries in both cases)

**Comment by seankhliao** [comment]
Is this still relevant to recent qemu/go releases?
